Use `recover()` to catch runtime panics

### Proposed Changes

- `net/http` has recover statement with defer in every goroutine it creates to catch panics
-  while running nuclei scan on 250 targets , it paniced after 3hrs 45 min 

```
[3:44:50] | Templates: 7384 | Hosts: 253 | RPS: 141 | Matched: 3255 | Errors: 50463 | Requests: 1911514/2797674 (68%)
panic: runtime error: slice bounds out of range [1721:49]

goroutine 26744483 [running]:
internal/poll.(*FD).Write(0x14016273600, {0x1403009ce00, 0x31, 0x32})
	internal/poll/fd_unix.go:380 +0x3ac
net.(*netFD).Write(0x14016273600, {0x1403009ce00?, 0x104f0f0e0?, 0x10512e220?})
	net/fd_posix.go:96 +0x28
net.(*conn).Write(0x14011ed7d48, {0x1403009ce00?, 0x14042d90fd8?, 0x10289b44c?})
	net/net.go:191 +0x34
github.com/miekg/dns.(*Conn).Write(0x1400dce5400, {0x1403009ce00, 0x31, 0x32})
	github.com/miekg/dns@v1.1.57/client.go:359 +0x108
github.com/miekg/dns.(*Conn).WriteMsg(0x1400dce5400, 0xc47f431c4ed?)
	github.com/miekg/dns@v1.1.57/client.go:348 +0xe4
github.com/miekg/dns.(*Client).ExchangeWithConnContext(0x140009faa80, {0x1051e4b78, 0x106722380}, 0x14015f11440, 0x1400dce5400)
	github.com/miekg/dns@v1.1.57/client.go:220 +0x2a8
github.com/miekg/dns.(*Client).ExchangeWithConn(...)
	github.com/miekg/dns@v1.1.57/client.go:187
github.com/miekg/dns.(*Client).Exchange(0x140451584e0?, 0x140451584e0?, {0x14013099540?, 0x104a08320?})
	github.com/miekg/dns@v1.1.57/client.go:170 +0xe0
github.com/projectdiscovery/retryabledns.(*Client).queryMultiple(0x14000b6d5f0, {0x140451584e0, 0x14}, {0x14042d913c4, 0x2, 0x1031545f0?}, {0x0, 0x0})
	github.com/projectdiscovery/retryabledns@v1.0.50/client.go:361 +0x718
github.com/projectdiscovery/retryabledns.(*Client).QueryMultiple(...)
	github.com/projectdiscovery/retryabledns@v1.0.50/client.go:264
github.com/projectdiscovery/retryabledns.(*Client).Resolve(0x1400039c6c0?, {0x140451584e0?, 0x140451584e0?})
	github.com/projectdiscovery/retryabledns@v1.0.50/client.go:144 +0x44
github.com/projectdiscovery/fastdialer/fastdialer.(*Dialer).GetDNSData(0x1400039c6c0, {0x140451584e0?, 0x1051c1700?})
	github.com/projectdiscovery/fastdialer@v0.0.52/fastdialer/dialer.go:489 +0x390
github.com/projectdiscovery/fastdialer/fastdialer.(*Dialer).dial(0x1400039c6c0, {0x1051e4da8, 0x14003ed2e00}, {0x103fcdf09, 0x3}, {0x140451584e0, 0x17}, 0x0, 0x0, 0x0, ...)
	github.com/projectdiscovery/fastdialer@v0.0.52/fastdialer/dialer.go:216 +0x29c
github.com/projectdiscovery/fastdialer/fastdialer.(*Dialer).Dial(...)
	github.com/projectdiscovery/fastdialer@v0.0.52/fastdialer/dialer.go:136
github.com/projectdiscovery/retryablehttp-go.DefaultReusePooledTransport.func1({0x1051e4da8?, 0x14003ed2e00?}, {0x103fcdf09?, 0x104d9dfe0?}, {0x140451584e0?, 0x1027f249c?})
	github.com/projectdiscovery/retryablehttp-go@v1.0.42/http.go:52 +0x64
net/http.(*Transport).dial(0x1400f4bb5c0?, {0x1051e4da8?, 0x14003ed2e00?}, {0x103fcdf09?, 0x14042d91a58?}, {0x140451584e0?, 0x1?})
	net/http/transport.go:1183 +0xdc
net/http.(*Transport).dialConn(0x14000c39180, {0x1051e4da8, 0x14003ed2e00}, {{}, 0x0, {0x14008fad780, 0x4}, {0x140451584e0, 0x17}, 0x0})
	net/http/transport.go:1625 +0x61c
net/http.(*Transport).dialConnFor(0x140253bf260?, 0x1401def9550)
	net/http/transport.go:1467 +0x7c
created by net/http.(*Transport).queueForDial in goroutine 26738384
	net/http/transport.go:1436 +0x380
```

this seems mostly due to open fd etc but i think maybe panic could have been avoided if we have added recover and caught it in fastdialer

--- a/fastdialer/dialer.go
+++ b/fastdialer/dialer.go
@@ -223,6 +223,11 @@ func (d *Dialer) DialZTLSWithConfig(ctx context.Context, network, address string
 }
 
 func (d *Dialer) dial(ctx context.Context, network, address string, shouldUseTLS, shouldUseZTLS bool, tlsconfig *tls.Config, ztlsconfig *ztls.Config, impersonateStrategy impersonate.Strategy, impersonateIdentity *impersonate.Identity) (conn net.Conn, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("panic: %v", r)
+		}
+	}()
 	var hostname, port, fixedIP string
 
 	if strings.HasPrefix(address, "[") {

--- a/fastdialer/dialer.go
+++ b/fastdialer/dialer.go
@@ -133,8 +133,8 @@ func NewDialer(options Options) (*Dialer, error) {
 		} else {
 			hostsFileData, err = metafiles.GetHostsFileDnsData(metafiles.Hybrid)
 		}
-		if err != nil {
-			return nil, err
+		if options.Logger != nil && err != nil {
+			options.Logger.Printf("could not load hosts file: %s\n", err)
 		}
 	}
 	dnsclient, err := retryabledns.New(resolvers, options.MaxRetries)

--- a/fastdialer/dialer.go
+++ b/fastdialer/dialer.go
@@ -127,10 +127,14 @@ func NewDialer(options Options) (*Dialer, error) {
 	var hostsFileData *hybrid.HybridMap
 	// load hardcoded values from host file
 	if options.HostsFile {
+		var err error
 		if options.CacheType == Memory {
-			hostsFileData, _ = metafiles.GetHostsFileDnsData(metafiles.InMemory)
+			hostsFileData, err = metafiles.GetHostsFileDnsData(metafiles.InMemory)
 		} else {
-			hostsFileData, _ = metafiles.GetHostsFileDnsData(metafiles.Hybrid)
+			hostsFileData, err = metafiles.GetHostsFileDnsData(metafiles.Hybrid)
+		}
+		if err != nil {
+			return nil, err
 		}
 	}
 	dnsclient, err := retryabledns.New(resolvers, options.MaxRetries)

--- a/fastdialer/dialer_test.go
+++ b/fastdialer/dialer_test.go
@@ -53,41 +53,41 @@ func testDialer(t *testing.T, options Options) {
 	fd.Close()
 }
 
-// nolint
-func testDialerIpv6(t *testing.T, options Options) {
-	// disk based
-	fd, err := NewDialer(options)
-	if err != nil {
-		t.Fatalf("couldn't create fastdialer instance: %s", err)
-	}
+// // nolint
+// func testDialerIpv6(t *testing.T, options Options) {
+// 	// disk based
+// 	fd, err := NewDialer(options)
+// 	if err != nil {
+// 		t.Fatalf("couldn't create fastdialer instance: %s", err)
+// 	}
 
-	// valid resolution + cache
-	ctx := context.Background()
-	conn, err := fd.Dial(ctx, "tcp", "ipv6.google.com:80")
-	if err != nil || conn == nil {
-		t.Fatalf("couldn't connect to target: %s", err)
-	}
-	conn.Close()
-	// retrieve cached data
-	data, err := fd.GetDNSData("ipv6.google.com")
-	if err != nil || data == nil {
-		t.Fatalf("couldn't retrieve dns data: %s", err)
-	}
-	if len(data.AAAA) == 0 {
-		t.Error("no AAAA results found")
-	}
+// 	// valid resolution + cache
+// 	ctx := context.Background()
+// 	conn, err := fd.Dial(ctx, "tcp", "ipv6.google.com:80")
+// 	if err != nil || conn == nil {
+// 		t.Fatalf("couldn't connect to target: %s", err)
+// 	}
+// 	conn.Close()
+// 	// retrieve cached data
+// 	data, err := fd.GetDNSData("ipv6.google.com")
+// 	if err != nil || data == nil {
+// 		t.Fatalf("couldn't retrieve dns data: %s", err)
+// 	}
+// 	if len(data.AAAA) == 0 {
+// 		t.Error("no AAAA results found")
+// 	}
 
-	// test address pinning
-	// this test passes, but will fail if the hard-coded ipv6 address changes
-	// need to find a better way to test this
-	/*
-		    conn, err = fd.Dial(ctx, "tcp", "ipv6.google.com:80:[2607:f8b0:4006:807::200e]")
-			if err != nil || conn == nil {
-				t.Errorf("couldn't connect to target: %s", err)
-			}
-			conn.Close()
-	*/
+// 	// test address pinning
+// 	// this test passes, but will fail if the hard-coded ipv6 address changes
+// 	// need to find a better way to test this
+// 	/*
+// 		    conn, err = fd.Dial(ctx, "tcp", "ipv6.google.com:80:[2607:f8b0:4006:807::200e]")
+// 			if err != nil || conn == nil {
+// 				t.Errorf("couldn't connect to target: %s", err)
+// 			}
+// 			conn.Close()
+// 	*/
 
-	// cleanup
-	fd.Close()
-}
+// 	// cleanup
+// 	fd.Close()
+// }

--- a/fastdialer/metafiles/shared.go
+++ b/fastdialer/metafiles/shared.go
@@ -39,51 +39,57 @@ func GetHostsFileDnsData(storage StorageType) (*hybrid.HybridMap, error) {
 	case Hybrid:
 		return getHFHybridStorage()
 	}
-	return nil, nil
+	return getHFInMemory()
 }
 
-var hostsMemOnce = &sync.Once{}
+// == initialize in-memory hostmap on demand
+var (
+	hmMem       *hybrid.HybridMap
+	hmErr       error
+	hostMemInit = sync.OnceFunc(func() {
+		opts := hybrid.DefaultMemoryOptions
+		hmMem, hmErr = hybrid.New(opts)
+		if hmErr != nil {
+			return
+		}
+		hmErr = loadHostsFile(hmMem, MaxHostsEntires)
+		if hmErr != nil {
+			hmMem.Close()
+			return
+		}
+	})
+)
 
 // getImm
 func getHFInMemory() (*hybrid.HybridMap, error) {
-	var hm *hybrid.HybridMap
-	var err error
-	hostsMemOnce.Do(func() {
-		opts := hybrid.DefaultMemoryOptions
-		hm, err = hybrid.New(opts)
-		if err != nil {
-			return
-		}
-		err = loadHostsFile(hm, MaxHostsEntires)
-		if err != nil {
-			hm.Close()
-			return
-		}
-	})
-	return hm, nil
+	hostMemInit()
+	return hmMem, hmErr
 }
 
-var hostsHybridOnce = &sync.Once{}
-
-func getHFHybridStorage() (*hybrid.HybridMap, error) {
-	var hm *hybrid.HybridMap
-	var err error
-	hostsHybridOnce.Do(func() {
+// == initialize hybrid hostmap on demand
+var (
+	hmHybrid  *hybrid.HybridMap
+	hmHybErr  error
+	hmHybInit = sync.OnceFunc(func() {
 		opts := hybrid.DefaultHybridOptions
 		opts.Cleanup = true
-		hm, err = hybrid.New(opts)
-		if err != nil {
+		hmHybrid, hmHybErr = hybrid.New(opts)
+		if hmHybErr != nil {
 			return
 		}
-		err = loadHostsFile(hm, -1)
-		if err != nil {
-			hm.Close()
+		hmHybErr = loadHostsFile(hmHybrid, -1)
+		if hmHybErr != nil {
+			hmHybrid.Close()
 			return
 		}
 		// set finalizer for cleanup
-		runtime.SetFinalizer(hm, func(hm *hybrid.HybridMap) {
+		runtime.SetFinalizer(hmHybrid, func(hm *hybrid.HybridMap) {
 			_ = hm.Close()
 		})
 	})
-	return hm, nil
+)
+
+func getHFHybridStorage() (*hybrid.HybridMap, error) {
+	hmHybInit()
+	return hmHybrid, hmHybErr
 }

--- a/fastdialer/options.go
+++ b/fastdialer/options.go
@@ -1,6 +1,7 @@
 package fastdialer
 
 import (
+	"log"
 	"net"
 	"time"
 
@@ -60,6 +61,7 @@ type Options struct {
 	OnDialCallback           func(hostname, IP string)
 	DisableZtlsFallback      bool
 	WithNetworkPolicyOptions *networkpolicy.Options
+	Logger                   *log.Logger // optional logger to log errors(like hostfile init error)
 }
 
 // DefaultOptions of the cache

